src/language-js/print/ternary.js

```diff
@@ -158,6 +160,79 @@ function printTernaryTest(path, options, print) {
   return printed;
 }
 
+const ancestorNameMap = new Map([
+  ["AssignmentExpression", "right"],
+  ["VariableDeclarator", "init"],
+  ["ReturnStatement", "argument"],
+  ["ThrowStatement", "argument"],
+  ["UnaryExpression", "argument"],
+  ["YieldExpression", "argument"],
+]);
+function shouldExtraIndentForConditionalExpression(path) {
+  const node = path.getValue();
+  if (node.type !== "ConditionalExpression") {
+    return false;
+  }
+
+  const parentName = path.callParent((path) => path.getName());
+  if (!parentName) {
+    return false;
+  }
+
+  let ancestorCount = 1;
+  let ancestor = path.getParentNode(ancestorCount);
+  while (isChainElement(ancestor)) {
+    ancestor = path.getParentNode(++ancestorCount);
+  }
+  const checkAncestor = (node) => {
+    const name = ancestorNameMap.get(ancestor.type);
+    return ancestor[name] === node;
+  };
+
+  const parent = path.getParentNode();
+
+  /**
+   * foo = (
+   *   condition
+   *     ? first
+   *     : second
+   * ) as SomeType;
+   */
+  if (parent.type === "TSAsExpression" && checkAncestor(parent)) {
```

Comment:
Let's merge `TSAsExpression` and `NewExpression` too, 

```js
  if (
    (
		(name === "callee" && parent.type === "NewExpression") ||
		(name === "expression" && parent.type === "TSAsExpression") 
	) &&
    checkAncestor(parent)
  )
```

Once again, `name === "expression"` not checked, `TSAsExpression` has two children, seems `typeAnnotation` can't be `ConditionalExpression`, but let's add it to be safer.